### ISSUE TITLE
Terraform crashes when locals try to access resources with count == 0

### ISSUE THREAD

**Issue by yves-vogl** [hashicorp/terraform#18608]
```
Terraform v0.11.7
+ provider.azurerm v1.12.0
+ provider.kubernetes v1.1.0
+ provider.local v1.1.0
+ provider.random v1.3.1
+ provider.tls v1.1.0
```

### Terraform Configuration Files

```hcl
(…)
module "app-1"  {
  source = "modules/app"

  resource_group_name = "${azurerm_resource_group.main.name}"
  location            = "${azurerm_resource_group.main.location}"

  name                = "app-1"
  project_name        = "${var.project_name}"
  maintainer          = "${var.maintainer}"

  # (...)
  disabled = true
}
module "app-2"  {
  source = "modules/app"

  resource_group_name = "${azurerm_resource_group.main.name}"
  location            = "${azurerm_resource_group.main.location}"

  name                = "app-2"
  project_name        = "${var.project_name}"
  maintainer          = "${var.maintainer}"

  # (...)
  disabled = true
}
# (…)

# in Module modules/app

resource "azurerm_container_registry" "main" {

  count               = "${var.disabled != "" ? 0 : 1}" # => count == 0 effectively

  resource_group_name = "${var.resource_group_name}"
  location            = "${var.location}"

  name                = "${random_id.project.hex}"
  # (…)
}



locals {
  dockercfg = {
    "${azurerm_container_registry.main.*.login_server}" = {
      username = "${azurerm_azuread_application.read.application_id}"
      password = "${var.container_registry_read_password}"
    }
  }
}
```

### Expected Behavior

Resources with count = 0 should not make locals crash.

### Actual Behavior
Resources with count = 0 are make locals crashing.

### Steps to Reproduce
Try to reference a resource with count = 0 through locals with splat definition.

See attached crash.log
[crash.log](https://github.com/hashicorp/terraform/files/2262837/crash.log)

**Comment by apparentlymart** [comment]
Hi @yves-vogl! Sorry for this crash, and thanks for reporting it.

I have a suspicion that the problem is somewhere other than it first appears:

```hcl
locals {
  dockercfg = {
    "${azurerm_container_registry.main.*.login_server}" = {
      username = "${azurerm_azuread_application.read.application_id}"
      password = "${var.container_registry_read_password}"
    }
  }
}
```

The `azurerm_container_registry.main.*.login_server` expression there will return an empty list when `count = 0`, and an empty list isn't a valid map key. While indeed it should not have _crashed_, the correct behavior here would be for Terraform to produce a type error that a string is required in that location.

I assume what you were hoping to do there is make `local.dockercfg` be an empty map when `count = 0`, and a single-element map when `count = 1`. In Terraform 0.11 there is no straightforward way to achieve that, but in the next major release [the conditional operator will work for all types](https://www.hashicorp.com/blog/terraform-0-12-conditional-operator-improvements) and we'll have [`for` expressions](https://www.hashicorp.com/blog/hashicorp-terraform-0-12-preview-for-and-for-each), which would allow the problem to be solved in one of the following two ways:

```hcl
# Example for forthcoming Terraform 0.12; details may change before release

locals {
  dockercfg = var.disabled ? {} : {
    "${azurerm_container_registry.main[0].login_server}" = {
      username = azurerm_azuread_application.read.application_id
      password = var.container_registry_read_password
    }
  }
}
```

```hcl
# Example for forthcoming Terraform 0.12; details may change before release

locals {
  dockercfg = {
    for x in azurerm_container_registry.main: x.login_server => {
      username = azurerm_azuread_application.read.application_id
      password = var.container_registry_read_password
    }
  }
}
```

For 0.11, I think the closest you could get here is to have the map contain an empty key when the repository is disabled:

```hcl
# Example for Terraform 0.11

locals {
  dockercfg = {
    "${join("", azurerm_container_registry.main.*.login_server)}" = {
      username = "${azurerm_azuread_application.read.application_id}"

      # Only export the password if the registry is enabled
      password = "${var.disabled ? "" : var.container_registry_read_password}"
    }
  }
}
```

**Comment by apparentlymart** [comment]
Hi again, @yves-vogl! Sorry for the long silence here.

We now have the changes I mentioned merged into master and so I was able to do some testing with the v0.12.0-alpha2 prerelease version.

First I verified that when a list is used as a map key it now returns a proper error rather than a crash:

```
Error: Incorrect key type

  on list-as-map-key.tf line 7, in locals:
   7:     "${null_resource.x.*.id}" = {
   8:       username = "foo"
   9:       password = "bar"
  10:     }

Can't use this value as a key: string required.
```

This shows that the crashing bug is fixed. :tada:

I also wanted to try the example I showed above of how this might be written using a `for` expression:

```hcl
locals {
  test = {
    for x in null_resource.x: x.id => {
      username = "foo"
      password = "bar"
    }
  }
}

output "test" {
  value = local.test
}
```

With `count = 0`:

```
$ terraform apply

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

test = {}
```

then with `count = 1`:

```
$ terraform apply

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # null_resource.x[0] will be created
  + resource "null_resource" "x" {
      + id = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

null_resource.x[0]: Creating...
null_resource.x[0]: Creation complete after 0s [id=5868995169152089727]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

Outputs:

test = {
  "5868995169152089727" = {
    "password" = "bar"
    "username" = "foo"
  }
}
```

This seems to be working as I expected!

Since these changes are already in master and ready to be included in the forthcoming v0.12.0 final release, I'm going to close this out now. Thanks again for reporting this, and thanks for your patience while we got to the bottom of it.
